pkg/storage/directory.go

```diff
@@ -0,0 +1,255 @@
+package storage
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/TierMobility/boring-registry/pkg/core"
+	"github.com/TierMobility/boring-registry/pkg/module"
+	"github.com/TierMobility/boring-registry/pkg/provider"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+var (
+	mirrorPrefix          = "mirror"
+	customProvidersPrefix = "providers"
+)
+
+// DirectoryStorage implements mirror.Storage
+type DirectoryStorage struct {
+	rwMutex sync.RWMutex
+	path    string
+}
+
+// EnumerateMirroredProviders stems from mirror.Storage
+func (d *DirectoryStorage) EnumerateMirroredProviders(ctx context.Context, provider core.Provider) (*[]core.Provider, error) {
+	return d.getProviders(ctx, mirrorPrefix, provider)
+}
+
+// RetrieveMirroredProviderArchive stems from mirror.Storage
+func (d *DirectoryStorage) RetrieveMirroredProviderArchive(ctx context.Context, provider core.Provider) (io.ReadCloser, error) {
+	fileName, err := provider.ArchiveFileName()
+	if err != nil {
+		return nil, err
+	}
+
+	f := fmt.Sprintf("%s/%s/%s/%s/%s/%s", d.path, mirrorPrefix, provider.Hostname, provider.Namespace, provider.Name, fileName)
+	file, err := os.Open(f)
+	if err != nil {
+		return nil, &ErrProviderNotMirrored{
+			Err:      err,
+			Provider: provider,
+		}
+	}
+
+	return io.NopCloser(bufio.NewReader(file)), nil
+}
+
+// StoreMirroredProvider stems from mirror.Storage
+func (d *DirectoryStorage) StoreMirroredProvider(ctx context.Context, provider core.Provider, binary, shasum, shasumSignature io.Reader) error {
+	// Acquiring lock, as the operation is not an atomic filesystem operation
```

Comment:
As the `directory` stoage backend is only experimental, the following is not going to be a real problem for us. But what happens if two goroutines want to store something simultaneously?
That's a good reason to support strongly consistent storage backends only